# SICP Chapter 2

In [1]:
;; from book -> not technically correct as will return negative
(define (gcd a b)
  (if (= b 0)
      a
      (gcd b (remainder a b))))

;; construct rational number reduced to lowest terms
(define (make-rat n d) 
  (let ((g (gcd n d)))
    (cons (/ n g)
          (/ d g))))

(define (numer x) (car x))
(define (denom x) (cdr x))

(define (add-rat x y)
  (make-rat (+ (* (numer x) (denom y))
               (* (numer y) (denom x)))
            (* (denom x) (denom y))))

(define (sub-rat x y)
  (make-rat (- (* (numer x) (denom y))
               (* (numer y) (denom x)))
            (* (denom x) (denom y))))

(define (mul-rat x y)
  (make-rat (* (numer x) (numer y))
            (* (denom x) (denom y))))

(define (div-rat x y)
  (make-rat (* (numer x) (denom y))
            (* (denom x) (numer y))))

(define (equal-rat? x y)
  (= (* (numer x) (denom y))
     (* (numer y) (denom x))))

(define (print-rat x)
  (newline)
  (display (numer x))
  (display "/")
  (display (denom x)))

In [2]:
(define half (make-rat 1 2))
(print-rat half)


1/2

In [3]:
(define fifth (make-rat 1 5))
(print-rat fifth)


1/5

In [4]:
(define third (make-rat 1 3))
(print-rat third)


1/3

In [5]:
(print-rat (mul-rat half fifth))


1/10

In [6]:
(print-rat (add-rat half fifth))


7/10

In [7]:
(print-rat (add-rat half half))


1/1

In [8]:
(print-rat (add-rat third third))


2/3

## 2.1

In [9]:
(define (make-rat n d) 
  (let ((g (abs (gcd n d)))) ; abs to correct for negative return from gcd
    (if (< d 0)
        (cons (/ (* n -1) g) 
              (/ (* d -1) g))
        (cons (/ n g) 
              (/ d g)))))
    

In [10]:
(print-rat (make-rat 1 2))


1/2

In [11]:
(print-rat (make-rat -1 2))


-1/2

In [12]:
(print-rat (make-rat 1 -2))


-1/2

## 2.2

In [13]:
(define (make-point x y)
  (cons x y))
(define (x-point point)
  (car point))
(define (y-point point)
  (cdr point))

(define (make-segment start-point end-point)
  (cons start-point end-point))
(define (start-segment segment)
  (car segment))
(define (end-segment segment)
  (cdr segment))

(define (midpoint-segment segment)
  (define (avg x y) (/ (+ x y) 2.0))
  (let ((start (start-segment segment)) (end (end-segment segment)))
    (make-point
     (avg (x-point start) (x-point end))
     (avg (y-point start) (y-point end)))))

(define (print-point p)
  (newline)
  (display "(")
  (display (x-point p))
  (display ",")
  (display (y-point p))
  (display ")"))

In [14]:
(print-point (make-point 1 2))


(1,2)

In [15]:
(define start (make-point 1 2))
(define end (make-point 2 4))

(define seg (make-segment start end))
(print-point (midpoint-segment seg))


(1.5,3.0)

## 2.3

In [16]:
;; opposite corners representation
(define (make-rect bottom-left top-right)
  (cons bottom-left top-right))
(define (rect-width rect)
  (abs (- (x-point (car rect)) (x-point (cdr rect)))))
(define (rect-height rect)
  (abs (- (y-point (car rect)) (y-point (cdr rect)))))

(define (rect-perim rect)
  (* 2 (+ (rect-width rect) (rect-height rect))))

(define (rect-area rect)
  (* (rect-width rect) (rect-height rect)))

In [17]:
(define rect (make-rect (make-point 1 2) (make-point 3 5)))

(rect-width rect)

2

In [18]:
(rect-height rect)

3

In [19]:
(rect-perim rect)

10

In [20]:
(rect-area rect)

6

In [21]:
;; bottom left corner + width/height representation
(define (make-rect bottom-left width height)
  (cons bottom-left (cons width height)))
(define (rect-width rect)
  (car (cdr rect)))
(define (rect-height rect)
  (cdr (cdr rect)))

In [22]:
(define rect (make-rect (make-point 1 2) 2 3))

In [23]:
(rect-perim rect)

10

In [66]:
(rect-area rect)

6

## 2.4

In [32]:
;; returns function that takes 2 arguments and applies it to x y
(define (cons x y)
  (lambda (m) (m x y)))

;; z takes a 2 argument function that is passed p q in order
;; returning first arg p implements car function
(define (car z)
  (z (lambda (p q) p)))

;; z takes a 2 argument function that is passed p q in order
;; returning sencond arg q produces cdr
(define (cdr z)
  (z (lambda (p q) q)))

In [27]:
(car (cons 1 2))

1

In [33]:
(cdr (cons 1 2))

2

Applicative-order evaluation to verify `(car (cons x y))` yields `x` for any objects `x` and `y`:
```scheme
(car (cons (x y)))
(car (lambda (m) (m x y)))
((lambda (m) (m x y)) (lambda (p q) p))
((lambda (p q) p) x y)
x
```

## 2.5

Represent pair of nonnegative integers $a, b$ as the integer that is the product $2^a3^b$

2 and 3 are prime -> $2^a3^b$ will give unique values for each all combinations of $a, b$ due to the [Fundamental Theorem of Arithmetic](http://mathworld.wolfram.com/FundamentalTheoremofArithmetic.html)

$a$ can be calculated from the product by the number of times it is evenly divisible by 2.

$b$ can be calculated from the product by the the number of times it is evenly divisible by 3.

Example:

$a=2\ ,b=3$

$2^a3^b=2^2\cdot3^3=108$

Calculating $a$:

$\frac{108}{2}=54$

$\frac{54}{2}=27$

$\frac{27}{2}=13.5$

2 even divisions by 2:

$a=2$

Calculating $b$:

$\frac{108}{3}=36$

$\frac{36}{3}=12$

$\frac{12}{3}=4$

$\frac{4}{3} = 1.3333333$

3 even divisions by 3:

$b=3$




In [11]:
(define (cons a b)
  (* (expt 2 a)
     (expt 3 b)))

; number of times n is evenly divisible by d
(define (num-div n d)
  (define (iter x count)
    (if (= 0 (remainder x d))
        (iter (/ x d) (+ 1 count))
        count))
  (iter n 0))

(define (car x)
  (num-div x 2))

(define (cdr x)
  (num-div x 3))

In [12]:
(car (cons 2 3))

2

In [13]:
(cdr (cons 2 3))

3

## 2.6

In [17]:
(define zero (lambda (f) (lambda (x) x)))

(define (add-1 n)
  (lambda (f) (lambda (x) (f ((n f) x)))))

Substitution to find definition of `one`:
```scheme
(add-1 zero)
(lambda (f) (lambda (x) (f ((zero f) x))))
(lambda (f) (lambda (x) (f ((lambda (x) x) x))))
(lambda (f) (lambda (x) (f x)))
```

In [19]:
;; applies given procedure once
(define one (lambda (f) (lambda (x) (f x))))

Substitution to find definition of `two`:
```scheme
(add-1 one)
(lambda (f) (lambda (x) (f ((one f) x))))
(lambda (f) (lambda (x) (f ((lambda (x) (f x)) x))))
(lambda (f) (lambda (x) (f (f x))))
```

In [20]:
;; applies given procedure twice
(define two (lambda (f) (lambda (x) (f (f x)))))

In [24]:
;; 'wrap' n in m function calls
(define (church+ m n)
  (lambda (f) (lambda (x) ((m f) ((n f) x)))))

In [38]:
;; easy to see number of function applications with increment procedure
(define (increment n)
  (+ n 1))

In [39]:
((zero increment) 0)

0

In [40]:
((zero increment) 1)

1

In [41]:
((zero increment) 2)

2

In [42]:
((one increment) 0)

1

In [43]:
((one increment) 1)

2

In [44]:
((one increment) 2)

3

In [45]:
((two increment) 0)

2

In [46]:
((two increment) 1)

3

In [47]:
((two increment) 2)

4

In [ ]:
(define three (church+ one two))

((three increment) 0)

## 2.7

In [1]:
;; procedures given in question description
(define (make-interval a b) (cons a b))

(define (add-interval x y)
  (make-interval (+ (lower-bound x)
                    (lower-bound y))
                 (+ (upper-bound x)
                    (upper-bound y))))

;; find minimum and maximum of products of the bounds
(define (mul-interval x y)
  (let ((p1 (* (lower-bound x)
               (lower-bound y)))
        (p2 (* (lower-bound x)
               (upper-bound y)))
        (p3 (* (upper-bound x)
               (lower-bound y)))
        (p4 (* (upper-bound x)
               (upper-bound y))))
    (make-interval (min p1 p2 p3 p4)
                   (max p1 p2 p3 p4))))
(define (div-interval x y)
  (mul-interval x
                (make-interval
                 (/ 1.0 (upper-bound y))
                 (/ 1.0 (lower-bound y)))))

In [2]:
(define (lower-bound x) (min (car x) (cdr x)))
(define (upper-bound x) (max (car x) (cdr x)))

In [3]:
;; test with a=lower, b=upper
(define x (make-interval 10 20))

(lower-bound x)

10

In [4]:
(upper-bound x)

20

In [5]:
;; test with a=upper, b=lower
(define y (make-interval 20 10))

(lower-bound y)

10

In [6]:
(upper-bound y)

20

## 2.8
Difference between intervals `x, y`:
- Minimum = lower bound `x` subtract upper bound `y` 
- Maximum = upper bound `x` subtract lower bound `y`

In [7]:
(define (sub-interval x y)
  (make-interval 
   (- (lower-bound x) (upper-bound y))
   (- (upper-bound x) (lower-bound y))))

(define x (make-interval 10 20))
(define y (make-interval 1 5))
(sub-interval x y)

(5 . 19)

## 2.9

`add-interval`:

$[a,b]+[c,d]=[(a+c), (b+d)]$

`width`:

$width([a,b])=\frac{b-a}{2}$

`width` of `add-interval`:

$width([a,b]+[c,d])=width([(a+c), (b+d)])$

$=\frac{(b+d)-(a+c)}{2}$

sum of `width` of two intervals:

$width([a,b]) + width([c,d])=\frac{b-a}{2}+\frac{d-c}{2}$

$=\frac{(b-a)+(d-c)}{2}$

$=\frac{(b+d)-(a+c)}{2}$

`width` of `add-interval` is equal to the sum of `with` of the same two intervals. Therefore the width of the the sum of two intervals is a function only of the widthds of the intervals being added.

In [8]:
(define (width x)
  (/ (- (upper-bound x) (lower-bound x)) 2))

(define a (make-interval 10 20))
(define b (make-interval 30 40))
(define c (make-interval 1 3))

In [9]:
(width a)

5

In [10]:
(width b)

5

In [11]:
(width c)

1

In [12]:
(width (add-interval a b))

10

Intervals `a` and `b` have the same width. Multiplying each of them by `c` gives produce intervals with different widths. Therefore the product of two intervals has a width that is not a function solely of the widths of said intervals.

In [13]:
(width (mul-interval a c))

25

In [14]:
(width (mul-interval b c))

45

## 2.10

In [15]:
(define (zero-span? x)
  (and (<= (lower-bound x) 0)
       (>= (upper-bound x) 0)))

;; throws an error if divide by 0
(define (div-interval x y)
  (if (zero-span? y)
      (error "div-interval" "Cannot divide by interval with span 0")
      (mul-interval x
                (make-interval
                 (/ 1.0 (upper-bound y))
                 (/ 1.0 (lower-bound y))))))

(div-interval (make-interval 10 20) (make-interval -1 1))


Traceback (most recent call last):
  File "In [15]", line 14, col 1, in 'div-interval'
  File "In [15]", line 8, col 7, in 'error'
  File "In [15]", line 8, col 7
RunTimeError: Error in 'div-interval': Cannot divide by interval with span 0



## 2.11
`(mul-interval x y)` produces an interval where the lower bound is the minimum product of `x` and `y`'s bounds  and the upper bound is the maximum product of `x` and `y`'s bounds.

Upper and lower bounds of an interval can be both positive, both negative or negative lower with positive upper i.e. `[+,+]`, `[-,-]` or `[-,+]`
- Can't have positive lower and negative upper as the lower bound would be less than the upper bound

For `mul-interval` this gives the combinations:
```
[+,+] * [+,+]
[+,+] * [-,-]
[+,+] * [-,+]

[-,-] * [-,-]
[-,-] * [+,+]
[-,-] * [-,+]

[-,+] * [-,+]
[-,+] * [+,+]
[-,+] * [-,-]
```

More than 2 multiplications are required for the case `[-,+] * [-,+]` as the product of the two negative lower bounds could be larger than that of the upper bounds.

In all other cases only two multiplications are required as there is only one combination of products that can produce the correct result
- e.g. `[+,+] * [+,+]` -> multiply two upper bounds and multiply two lower bounds.

In [16]:
(define (pos? x) (>= x 0))
(define (neg? x ) (<= x 0))

(define (mul-interval x y)
  (let ((lx (lower-bound x))
        (ux (upper-bound x))
        (ly (lower-bound y))
        (uy (upper-bound y)))
    (cond ((and (pos? lx)
                (pos? ux)
                (pos? ly)
                (pos? uy))
           ;; [+, +] * [+, +]
           (make-interval (* lx ly) (* ux uy)))
          ((and (pos? lx)
                (pos? ux)
                (neg? ly)
                (pos? uy))
           ;; [+, +] * [-, +]
           (make-interval (* ux lx) (* ux uy)))
          ((and (pos? lx)
                (pos? ux)
                (neg? ly)
                (neg? uy))
           ;; [+, +] * [-, -]
           (make-interval (* ux lx) (* lx uy)))
          ((and (neg? lx)
                (pos? ux)
                (pos? ly)
                (pos? uy))
           ;; [-, +] * [+, +]
           (make-interval (* lx uy) (* ux uy)))
          ((and (neg? lx)
                (pos? ux)
                (neg? ly)
                (pos? uy))
           ;; [-, +] * [-, +]
           (make-interval (min (* ux ly) (* lx uy))
                          (max (* lx ly) (* ux uy))))
          ((and (neg? lx)
                (pos? ux)
                (neg? ly)
                (neg? uy))
           ;; [-, +] * [-, -]
           (make-interval (* ux ly) (* lx ly)))
          ((and (neg? lx)
                (neg? ux)
                (pos? ly)
                (pos? uy))
           ;; [-, -] * [+, +]
           (make-interval (* lx ux) (* ux ly)))
          ((and (neg? lx)
                (neg? ux)
                (neg? ly)
                (pos? uy))
           ;; [-, -] * [-, +]
           (make-interval (* lx ux) (* lx ly)))
          ((and (neg? lx)
                (neg? ux)
                (neg? ly)
                (neg? uy))
           ;; [-, -] * [-, -]
           (make-interval (* ux uy) (* lx ly))))))

In [17]:
(define a (make-interval 1 2))
(define b (make-interval -1 2))
(define c (make-interval -1 -2))

In [18]:
(mul-interval a a)

(1 . 4)

In [19]:
(mul-interval a b)

(2 . 4)

In [20]:
(mul-interval a c)

(2 . -1)

In [21]:
(mul-interval b a)

(-2 . 4)

In [22]:
(mul-interval b b)

(-2 . 4)

In [23]:
(mul-interval b c)

(-4 . 2)

In [24]:
(mul-interval c a)

(2 . -1)

In [25]:
(mul-interval c b)

(2 . 2)

In [65]:
(mul-interval c c)

(1 . 4)

## 2.12

In [27]:
;; definitions given in question
(define (make-center-width c w)
  (make-interval (- c w) (+ c w)))

(define (center i)
  (/ (+ (lower-bound i)
        (upper-bound i))
     2))

(define (width i)
  (/ (- (upper-bound i)
        (lower-bound i))
     2))

In [62]:
;; using make-interval directly
(define (make-center-percent c tolerance)
  (let ((unc (* (/ c 100) tolerance)))
    (make-interval (- c unc) (+ c unc))))

;; using provided make-center-width
(define (make-center-percent c tolerance)
  (make-center-width c (* (/ c 100) tolerance)))

(define (percent i)
  (* (/ (width i) (center i)) 100.0))

In [63]:
(define i (make-center-percent 10 25))
i

(15/2 . 25/2)

In [64]:
(center i)

10

In [65]:
(percent i)

25.0

## 2.13

Interval multiplication:

$[a,b]\cdot[c,d] = [\min(ac,ad,bc,bd), \max(ac,ad,bc,bd)]$

Assuming all values are positive:

$[a,b]\cdot[c,d] = [ac, bd]$

Interval represented by its center $c_i$ and tolerance percentage $p_i$ :

$i = [c_i-\frac{c_i\cdot p_i}{100},c_i+\frac{c_i\cdot p_i}{100}]$

$= [c_i\frac{1-p_i}{100},c_i\frac{1+p_i}{100}]$

Multiplying two intervals $i,j$:

$ij=[c_i\frac{1-p_i}{100},c_i\frac{1+p_i}{100}]\cdot[c_j\frac{1-p_j}{100},c_j\frac{1+p_j}{100}]$

$=[c_i\frac{1-p_i}{100}\cdot c_j\frac{1-p_j}{100}, c_i\frac{1+p_i}{100}\cdot c_j\frac{1+p_j}{100}]$

$=[c_ic_j(1-\frac{p_i+p_j}{100}+\frac{p_ip_j}{10000}), c_ic_j(1+\frac{p_i+p_j}{100}+\frac{p_ip_j}{10000})]$

Percentage tolerances are small ->$\frac{p_ip_j}{10000}$ terms will be small -> approximate as:

$ij=[c_ic_j(1-\frac{p_i+p_j}{100}), c_ic_j(1+\frac{p_i+p_j}{100})]$

Percentage tolerance of the product = $p_i+p_j$
- **sum of the tolerances of the two intervals**

In [67]:
;; test with small tolerances
(define i (make-center-percent 100 2))
(define j (make-center-percent 80 1))

(define ij (mul-interval i j))
;; approximately 2+1=3
(percent ij)

2.9994001199760048

## 2.14

In [68]:
;; procedures from question
(define (par1 r1 r2)
  (div-interval
   (mul-interval r1 r2)
   (add-interval r1 r2)))

(define (par2 r1 r2)
  (let ((one (make-interval 1 1)))
    (div-interval
     one
     (add-interval 
      (div-interval one r1)
      (div-interval one r2)))))

In [85]:
;; demonstrate different results between par1 and par2
(define A (make-center-percent 100 3))
(define B (make-center-percent 200 2))

(par1 A B)

(61.928338762214985 . 71.71331058020478)

In [86]:
(par2 A B)

(64.88737201365187 . 68.44299674267101)

In [87]:
;; investigate behaviour with arithmetic expressions
(define A/A (div-interval A A))
(define A/B (div-interval A B))

(center A/A)

1.0018016214593133

In [88]:
(percent A/A)

5.994604855629938

In [89]:
(center A/B)

0.500500200080032

In [90]:
(percent A/B)

4.997001798920647

The expected center of dividing an interval by itself is 1, however in the investigation above, `(center A/A)` returns `1.0018016214593133` - an approximation to 1.

## 2.15
The question states that `par1` and `par2` use two different but algebraically equivalent formulae:
- `par1` uses the formula $\frac{R_1R_2}{R_1+R_2}$
- `par2` uses the formula $\frac{1}{\frac{1}{R_1}+\frac{1}{R_2}}$

Proof of algrebraic equivalence:

Multiply by $\frac{R_1}{R_1}$ and $\frac{R_2}{R_2}$. These are both equivalent to 1 and so the formula is still valid:

$\frac{1}{\frac{1}{R_1}+\frac{1}{R_2}}=\frac{R_1}{R_1}\frac{R_2}{R_2}\frac{R_1R_2}{R_1+R_2}$

$=\frac{R_1R_2}{R_1R_2(\frac{1}{R_1}+\frac{1}{R_2})}$

$=\frac{R_1R_2}{\frac{R_1R_2}{R_1}+\frac{R_1R_2}{R_2}}$

$=\frac{R_1R_2}{R_1+R_2}$

The transformation assumes that $\frac{R_1}{R_1}$ and $\frac{R_2}{R_2}$ equal 1. However it was shown in question 2.14 that dividing an interval by itself will produce a result with a center that is an **approximation** to 1 - introducing the descrepancy between `par1` and `par2`. Therefore, tighter error bounds will be achieved by avoiding repetition of variables that represent uncertain numbers - **Eva Lu Ator is correct**.

## 2.17

In [1]:
(define (last-pair l)
  (if (null? (cdr l))
      l
      (last-pair (cdr l))))

In [2]:
(last-pair (list 23 72 149 34))

(34)

## 2.18

In [79]:
;; given in question
(define (append list1 list2)
  (if (null? list1)
      list2
      (cons (car list1)
            (append (cdr list1)
                    list2))))

;; reverse by appending car of a list to the reverse of cdr of the list
(define (reverse l)
  (if (null? l)
      l
      (append (reverse (cdr l)) (list (car l)))))

(reverse (list 1 4 9 16 25))

(25 16 9 4 1)

## 2.19

In [3]:
;; given in question
(define us-coins 
  (list 50 25 10 5 1))

(define (cc amount coin-values)
  (cond ((= amount 0)
         1)
        ((or (< amount 0)
             (no-more? coin-values))
         0)
        (else
         (+ (cc
             amount
             (except-first-denomination
              coin-values))
            (cc
             (- amount
                (first-denomination
                 coin-values))
             coin-values)))))

(define (first-denomination coin-values)
  (car coin-values))

(define (except-first-denomination coin-values)
  (cdr coin-values))

(define (no-more? coin-values)
  (null? coin-values))

In [27]:
(cc 100 us-coins)

292

##### The order of the list `coin-values` will **not effect** the answer produced by `cc`
- Each sub-list of `coin-values` is evaluated recursively after subtracting the value of the first coin from the amount
- All possible combinations are computed regardless of the order of `coin-values`

In [4]:
(cc 100 (reverse us-coins))

292

In [6]:
(define us-coins-shuffled (list 25 5 50 1 10))
(cc 100 us-coins-shuffled)

292

## 2.20

In [21]:
;; check parity of first element x
;; filter list using function that determines if input is that parity
(define (same-parity x . y)
  (define (iter z filter)
    (cond ((null? z) z)
          ((filter (car z))
            ;; parity matches append first item, check next
           (cons (car z) (iter (cdr z) filter)))
          ;; parity mismatch, check next
          (else (iter (cdr z) filter))))
  (iter (cons x y) (if (even? x) even? odd?)))

In [22]:
(same-parity 1 2 3 4 5 6 7)

(1 3 5 7)

In [23]:
(same-parity 2 3 4 5 6 7)

(2 4 6)

## 2.21

In [22]:
(define (map proc items)
  (if (null? items)
      items
      (cons (proc (car items))
            (map proc (cdr items)))))

(define (square-list items)
  (if (null? items)
      items
      (cons (* (car items) (car items)) 
            (square-list (cdr items)))))

(square-list (list 1 2 3 4))

(1 4 9 16)

In [23]:
(define (square-list items)
  (map (lambda (x) (* x x)) items))

In [24]:
(square-list (list 1 2 3 4))

(1 4 9 16)

## 2.22

In [38]:
(define (square x) (* x x))

;; Result list is reversed
(define (square-list items)
  (define (iter things answer)
    (if (null? things)
        answer
        (iter (cdr things)
              ;; square item from front of input list and append to 
              ;; front of answer list
              (cons (square (car things))
                    answer))))
  ;; start with empty list due to implementation details of nil value
  (iter items (list)))

(square-list (list 1 2 3 4))

(16 9 4 1)

On each iteration, an item is taken from the front of the original list, is squared and then appends the result to the *front* of the answer list. The last item to be squared from the original list will therfore end up as the first item in the answer list 

In [39]:
(define (square-list items)
  (define (iter things answer)
    (if (null? things)
        answer
        (iter (cdr things)
              ;; pass list as first arg to cons
              (cons answer 
                    (square
                     (car things))))))
  (iter items (list)))

(square-list (list 1 2 3 4))

((((() . 1) . 4) . 9) . 16)

Flipping the order of arguments to `cons` creates a pair containing the answer list and the next squared integer. It does *not* append the integer onto the list. Each iteration combines the list from the previous step with the next integer result and wraps it in a new list.

## 2.23

In [44]:
(define (for-each proc items)
  (cond ((null? items)
         #t)
        (else (proc (car items))
              (for-each proc (cdr items)))))

(for-each 
 (lambda (x) (newline) (display x))
 (list 1 2 3 4))


1
2
3
4

#t

## 2.24 TODO - ADD DIAGRAMS

In [45]:
(list 1 (list 2 (list 3 4)))

(1 (2 (3 4)))

## 2.25

In [47]:
(define a (list 1 3 (list 5 7) 9))
a

(1 3 (5 7) 9)

In [57]:
(car (cdr (car (cdr (cdr a)))))

7

In [48]:
(define b (list (list 7)))
b

((7))

In [56]:
(car (car b))

7

In [49]:
(define c (list 1 (list 2 (list 3 (list 4 (list 5 (list 6 7)))))))
c

(1 (2 (3 (4 (5 (6 7))))))

In [72]:
(cdr (car (cdr (car (cdr (car (cdr (car (cdr (car (cdr c)))))))))))

(7)

## 2.26

In [73]:
(define x (list 1 2 3))
(define y (list 4 5 6))

In [76]:
(append x y)

(1 2 3 4 5 6)

In [77]:
(cons x y)

((1 2 3) 4 5 6)

In [78]:
(list x y)

((1 2 3) (4 5 6))

## 2.27

In [83]:
;; from 2.18
(define (reverse l)
  (if (null? l)
      l
      (append (reverse (cdr l)) (list (car l)))))

(define (deep-reverse l)
  (cond ((null? l) (list))
        ((pair? (car l))
         (append (deep-reverse (cdr l))
                 (list (deep-reverse (car l)))))
        (else (append (deep-reverse (cdr l))
                      (list (car l))))))

In [84]:
(define x (list (list 1 2) (list 3 4)))
x

((1 2) (3 4))

In [85]:
(reverse x)

((3 4) (1 2))

In [88]:
(deep-reverse x)

((4 3) (2 1))

## 2.28

In [103]:
;; traverse tree->append each node to list result,
(define (fringe t)
  (cond ((null? t) (list))
        ((not (pair? t)) (list t))
        (else (append (fringe (car t))
                      (fringe (cdr t))))))

In [104]:
(define x (list (list 1 2) (list 3 4)))
x

((1 2) (3 4))

In [105]:
(fringe x)

(1 2 3 4)

## 2.29

In [111]:
(define (make-mobile left right)
  (list left right))

(define (make-branch len structure)
  (list len structure))

(define (left-branch mobile)
  (car mobile))

(define (right-branch mobile)
  (car (cdr mobile)))

(define (branch-length branch)
  (car branch))

(define (branch-structure branch)
  (car (cdr branch)))

;; mobile total weight = sum of left + right branch weight
;; branch weight = value of weight if not a mobile, or recursive if branch
;; holds a mobile
(define (branch-weight branch)
  (if (pair? (branch-structure branch))
      (total-weight (branch-structure branch))
      (branch-structure branch)))

(define (total-weight mobile)
  (+ (branch-weight (left-branch mobile))
     (branch-weight (right-branch mobile))))

(define (branch-torque branch)
  (* (branch-length branch)
     (branch-weight branch)))

(define (branch-balanced? branch)
  (if (pair? (branch-structure branch))
      (mobile-balanced? (branch-structure branch))
      #t))
(define (mobile-balanced?)
  (and (= (branch-torque (left-branch mobile))
          (branch-torque (right-branch mobile)))
       (branch-balanced? (left-branch mobile))
       (branch-balanced? (right-branch mobile))))

Changing the representation to:
```scheme
(define (make-mobile left right)
  (cons left right))

(define (make-branch length structure)
  (cons length structure))
```

Would only affect the `right-branch` and `branch-structure` accessors:
```scheme
(define (right-branch mobile)
  (cdr mobile))

(define (branch-structure branch)
  (cdr branch))
```

## 2.30

In [5]:
(define t (list 1
                (list 2 (list 3 4) 5)
                (list 6 7)))
t

(1 (2 (3 4) 5) (6 7))

In [6]:
(define (square-tree tree)
  (cond ((null? tree) (list))
        ((not (pair? tree))
         (* tree tree))
        (else
         (cons (square-tree (car tree))
               (square-tree (cdr tree))))))

In [7]:
(square-tree t)

(1 (4 (9 16) 25) (36 49))

In [8]:
(define (square-tree tree)
  (map (lambda (sub-tree)
         (if (pair? sub-tree)
             (square-tree sub-tree)
             (* sub-tree sub-tree)))
       tree))

In [9]:
(square-tree t)

(1 (4 (9 16) 25) (36 49))

## 2.31

In [12]:
(define (tree-map proc tree)
  (map (lambda (sub-tree)
         (if (pair? sub-tree)
             (tree-map proc sub-tree)
             (proc sub-tree)))
       tree))

(define (square x) (* x x))

(define (square-tree tree)
  (tree-map square tree))

(square-tree t)

(1 (4 (9 16) 25) (36 49))

## 2.32
Set of subsets of a set = **powerset**

Algorithm from [Wikipedia](https://en.wikipedia.org/wiki/Power_set#Algorithms):
- Powerset of empty set = the set containing the empty set
- Powerset of any other set is all the subsets of the set containing a specific element and all the subsets of the set not containing that element


$S=\{1, 2, 3\}$

$P(S)=\{\{\}\{3\}\{2\}\{2,3\}\{1\}\{1,3\}\{1,2\}\{1,2,3\}\}$

In [36]:
(define (subsets s)
  (if (null? s)
      (list ())
      (let ((rest (subsets (cdr s))))
        (append rest (map (lambda (x) (cons (car s) x)) rest)))))

In [37]:
(subsets (list 1 2 3))

(() (3) (2) (2 3) (1) (1 3) (1 2) (1 2 3))

Line `3` of `subsets` represents the base case:
- $P(\{\})=\{\{\}\}$
`subsets` is then called recursively with `(cdr s)`- is all but the first element of the set. The first element of the set - `(car s)` - is then appended to each subset of `(cdr s)`. The recursion stops when `(cdr s)` on line `4` of `subsets` has no elements and the base case is reached.

## 2.33

In [2]:
;; accumulate definition from book
(define (accumulate op initial sequence)
  (if (null? sequence)
      initial
      (op (car sequence)
          (accumulate op
                      initial
                      (cdr sequence)))))

In [3]:
(accumulate + 0 (list 1 2 3))

6

In [6]:
(accumulate * 1 (list 4 5 6))

120

In [10]:
(define (map p sequence)
  (accumulate (lambda (x y) (cons (p x) y))
              (list) sequence))

(define (square x) (* x x))
(map square (list 1 2 3))

(1 4 9)

In [12]:
(define (append seq1 seq2)
  (accumulate cons seq2 seq1))

(append (list 1 2 3) (list 4 5 6))

(1 2 3 4 5 6)

In [15]:
(define (length sequence)
  (accumulate (lambda (x y) (+ y 1)) 0 sequence))

(length (list 1 2 3))

3

## 2.34

### Horner's rule
Evaluation of the polynomial:

$a_nx^n+a_{n-1}x^{n-1}+...+a_1x+a_0$

At a given value of $x$ can be computed as:

$(...(a_nx+a_{n-1})x+...+a_1)x+a_0$

In [18]:
(define
  (horner-eval x coefficient-sequence)
  (accumulate
   (lambda (this-coeff higher-terms)
     (+ (* x higher-terms) this-coeff))
   0
  coefficient-sequence))

In [19]:
(horner-eval 2 (list 1 3 0 5 0 1))

79

## 2.35

In [4]:
;; from book
(define (enumerate-tree tree)
  (cond ((null? tree) (list))
         ((not (pair? tree)) (list tree))
         (else (append
                (enumerate-tree (car tree))
                (enumerate-tree (cdr tree))))))

(define (count-leaves t)
  (accumulate + 0 (map (lambda (x) 1)
                       (enumerate-tree t))))

In [5]:
(count-leaves (cons (list 1 2) (list 3 4)))

4

## 2.36

In [6]:
(define (accumulate-n op init seqs)
  (if (null? (car seqs))
      (list)
      (cons (accumulate op init (map car seqs))
            (accumulate-n op init (map cdr seqs)))))

In [7]:
(define s (list (list 1 2 3)
                (list 4 5 6) 
                (list 7 8 9) 
                (list 10 11 12)))
s

((1 2 3) (4 5 6) (7 8 9) (10 11 12))

In [51]:
(accumulate-n + 0 s)

(22 26 30)

## 2.37

In [1]:
;; representation of a vector
(list 1 2 3 4)

(1 2 3 4)

In [4]:
;; representation of a matrix
(define m (list (list 1 2 3 4)
                (list 4 5 6 6)
                (list 6 7 8 9)))
m

((1 2 3 4) (4 5 6 6) (6 7 8 9))

In [15]:
;; helper function to print a matrix
(define (print-matrix m)
  (map (lambda (row) (display row)(newline)) m))

(print-matrix m)

(1 2 3 4)
(4 5 6 6)
(6 7 8 9)


(<void> <void> <void>)

In [9]:
;; given in book
(define (dot-product v w)
  (accumulate + 0 (map * v w)))

(dot-product (list 1 2 3 4) (list 4 5 6 7))

60

In [10]:
(define (matrix-*-vector m v)
  (map (lambda (row) (dot-product row v)) m))

(matrix-*-vector m (list 1 2 3 4))

(30 56 80)

In [13]:
;; make columns of input matrix rows of output matrix
(define (transpose mat)
  (accumulate-n cons (list) mat))

(print-matrix (transpose m))

(1 4 6)
(2 5 7)
(3 6 8)
(4 6 9)


(<void> <void> <void> <void>)

In [20]:
(define (matrix-*-matrix m n)
  (let ((cols (transpose n)))
    (map (lambda (row) (matrix-*-vector cols row)) m)))

(define n (list (list 1 2 5 8)
                (list 9 5 2 1)
                (list 8 7 3 4)))

(display "m:")(newline)
(print-matrix m)
(display "n:")(newline)
(print-matrix n)
(display "m*n:")(newline)
(print-matrix (matrix-*-matrix m n))

m:
(1 2 3 4)
(4 5 6 6)
(6 7 8 9)
n:
(1 2 5 8)
(9 5 2 1)
(8 7 3 4)
m*n:
(43 33 18 22)
(97 75 48 61)
(133 103 68 87)


(<void> <void> <void>)

## 2.38
```
3/2
1/6
(1 (2 (3 ())))
(((() 1) 2) 3)
```

In [8]:
;; same as accumulate
(define (fold-right op initial sequence)
   (if (null? sequence)
       initial
       (op (car sequence)
           (fold-right op initial (cdr sequence)))))

(define (fold-left op initial sequence)
  (define (iter result rest)
    (if (null? rest)
        result
        (iter (op result (car rest))
              (cdr rest))))
  (iter initial sequence))

In [29]:
(fold-right / 1 (list 1 2 3))

3/2

In [30]:
(fold-left / 1 (list 1 2 3))

1/6

In [31]:
(fold-right list (list) (list 1 2 3))

(1 (2 (3 ())))

In [32]:
(fold-left list (list) (list 1 2 3))

(((() 1) 2) 3)

`op` should be **commutative** for `fold-left` and `fold-right`to produce the same values for any sequence.

The two procedures differ only in the order in which they process the sequence elements. This means that if `op` is not commutative, different results will be produced.

`*` is commutative, therefore `fold-left` and `fold-right` produce the same values:

In [11]:
(fold-right * 1 (list 1 2 3))

6

In [12]:
(fold-left * 1 (list 1 2 3))

6

`/` is *not* commutative, therefore `fold-left` and `fold-right` prduce different values:

In [13]:
(fold-right / 1 (list 1 2 3))

3/2

In [14]:
(fold-left / 1 (list 1 2 3))

1/6

## 2.39

In [9]:
(define (reverse sequence)
  (fold-right
  (lambda (x y) (append y (list x))) (list) sequence))

In [40]:
(reverse (list 1 2 3))

(3 2 1)

In [69]:
(define (reverse sequence)
  (fold-left
   (lambda (x y) (cons y x)) (list) sequence))

In [70]:
(reverse (list 1 2 3))

(3 2 1)

## 2.40

In [40]:
;; calysto scheme missing filter function
(define (filter predicate seq)
  (accumulate (lambda (e rest)
                (if (predicate e)
                    (cons e rest)
                    rest)) 
              ()
              seq))

(define (enumerate-interval low high)
  (if (> low high)
      ()
      (cons low 
            (enumerate-interval
             (+ low 1)
             high))))

(define (flatmap proc seq)
  (accumulate append () (map proc seq)))

(define (prime-sum? pair)
  (prime? (+ (car pair) (cadr pair))))

(define (make-pair-sum pair)
  (list (car pair)
        (cadr pair)
        (+ (car pair) (cadr pair))))

(define (unique-pairs n)
  (flatmap
   (lambda (i)
     (map (lambda (j) (list i j))
          (enumerate-interval 1 (- i 1))))
   (enumerate-interval 1 n)))

(unique-pairs 6)

((2 1) (3 1) (3 2) (4 1) (4 2) (4 3) (5 1) (5 2) (5 3) (5 4) (6 1) (6 2) (6 3) (6 4) (6 5))

In [41]:
(define (prime-sum-pairs n)
  (map make-pair-sum
       (filter prime-sum? (unique-pairs n))))

## 2.41

In [42]:
;; produce ordered triples of distinct positive integers (i,j,k) up to n
(define (ordered-triples n)
  (flatmap (lambda (i)
             (flatmap (lambda (j)
                        (map (lambda (k) (list i j k))
                             (enumerate-interval 1 (- j 1))))
                      (enumerate-interval 1 (- i 1))))
           (enumerate-interval 1 n)))

(ordered-triples 6)

((3 2 1) (4 2 1) (4 3 1) (4 3 2) (5 2 1) (5 3 1) (5 3 2) (5 4 1) (5 4 2) (5 4 3) (6 2 1) (6 3 1) (6 3 2) (6 4 1) (6 4 2) (6 4 3) (6 5 1) (6 5 2) (6 5 3) (6 5 4))

In [43]:
;; true if sum of triple = s
(define (triple-sum? triple s)
  (= (accumulate + 0 triple) s))

(triple-sum? (list 1 2 3) 6)

#t

In [44]:
(triple-sum? (list 1 2 3) 5)

#f

In [45]:
(define (ordered-triple-sum n s)
  (filter (lambda (triple) (triple-sum? triple s)) (ordered-triples n)))

In [46]:
(ordered-triple-sum 5 10)

((5 3 2) (5 4 1))

In [47]:
(ordered-triple-sum 10 5)

()

In [51]:
(ordered-triple-sum 10 6)

((3 2 1))

In [52]:
(ordered-triple-sum 20 8)

((4 3 1) (5 2 1))